src/hierarchical.rs

```diff
@@ -0,0 +1,244 @@
+use std::cmp::{min, max};
+use std::f64::{INFINITY, NEG_INFINITY};
+
+use {Point};
+
+// TODO: Add tests
+// TODO: More efficient `Elements` Iterator
+// TODO: Implement `SLINK` as single-linkage criterion
+// TODO: Implement `CLINK` as alternative complete-linkage criterion
+
+
+/// Hierarchical linkage criteria.
+///
+/// The linkage decides how the distance between clusters is computed.
+///
+/// <script src="https://is.gd/BFouBe"></script>
+#[derive(Copy, Clone, Debug)]
+pub enum LinkageCriterion {
+    /// Maximum or complete-linkage clustering.
+    ///
+    /// <p>
+    /// $$\max_{a \in A,\, b \in B} \, d(a, b)$$
+    /// </p>
+    Complete,
+
+    /// Minimum or single-linkage clustering.
+    ///
+    /// <p>
+    /// $$\min_{a \in A,\, b \in B} \, d(a, b)$$
+    /// </p>
+    Single
+}
+
+
+/// Agglomerative clustering, i.e., hierarchical bottum-up clustering.
+///
+/// Starting with each element in its own cluster, iteratively merge the two most similar
+/// clusters until a certain distance threshold is reached. The similarity between two
+/// clusters is mainly defined by the linkage criterion and the `Point`'s distance
+/// metric.
+///
+/// The inter-point distances are pre-computed and cached in a distance matrix.
+///
+/// * **Complete-linkage**: Currently a naive implementation with O( n^3 ) time
```

Comment:
These naive implementations can be swapped out for the optimal ones without changing the API, right?

(Have you looked into how difficult implementing the faster ones is? I took a glance at the CLINK paper, but not enough time to digest it in any nontrivial way: it seems rather dense!)